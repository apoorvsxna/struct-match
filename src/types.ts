import { SyntaxNode } from 'tree-sitter';

/** An alias for a tree-sitter `SyntaxNode`. */
export interface ASTNode extends SyntaxNode {}

/** A map of captured metavariable names to their string values. */
export interface Metavariable {
  [key: string]: string;
}

/** A single match result from a search query. */
export interface Match {
  /** The id of the rule that produced this match. */
  ruleId: string;
  /** The AST node that matched. */
  node: ASTNode;
  /** The text content of the matched node. */
  matchedText: string;
  startLine: number;
  endLine: number;
  /** Captured metavariables (e.g., `{ "$NAME": "user" }`). */
  metavariables: Metavariable;
}

/**
 * A declarative rule for querying the AST.
 * Rules can be nested to specify context.
 */
export interface RulePattern {
  /** The core tree-sitter pattern string. */
  pattern?: string;
  /** Matches any of the provided patterns. */
  any?: RulePattern[];
  /** Narrows the search to descendants of a node matching this pattern. */
  inside?: RulePattern;
  /** Requires the matched node to contain a node matching this pattern. */
  contains?: RulePattern;
  /** Requires a preceding sibling to match this pattern. */
  precedes?: RulePattern;
  /** Requires a following sibling to match this pattern. */
  follows?: RulePattern;
  /** Inverts the match; excludes nodes that match this pattern. */
  not?: RulePattern;
}