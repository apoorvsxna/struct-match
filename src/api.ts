import { Match, RulePattern } from './types';
import { CodeParser } from './core/Parser';
import { Matcher } from './core/Matcher';
import { RuleProcessor } from './core/RuleProcessor';
import { getLineNumbers, getMatchText } from './utils/NodeAnalyzer';
import yaml from 'js-yaml';

const codeParser = new CodeParser();
const matcher = new Matcher(codeParser);
const ruleProcessor = new RuleProcessor(matcher);

/**
 * Finds all occurrences of a simple pattern string in the given source code.
 *
 * @param sourceCode The source code to analyze.
 * @param pattern The pattern to search for (e.g., "console.log($VAR)").
 * @returns A promise that resolves to an array of all found matches.
 */
export async function findByPattern(
  sourceCode: string,
  pattern: string
): Promise<Omit<Match, 'ruleId'>[]> { // <-- The fix is here: added []
  const sourceTree = codeParser.parse(sourceCode);
  const rule: RulePattern = { pattern };

  const intermediateMatches = matcher.find(sourceTree, rule);

  return intermediateMatches.map(({ node, metavariables }) => {
    const { startLine, endLine } = getLineNumbers(sourceCode, node);
    const matchedText = getMatchText(node);

    return {
      node,
      matchedText,
      startLine,
      endLine,
      metavariables,
    };
  });
}

/**
 * Finds all matches in the source code based on a structured YAML rule.
 *
 * @param sourceCode The source code to analyze.
 * @param yamlRule A string containing the rule (with id) in YAML format.
 * @returns A promise that resolves to an array of all found matches.
 */
export async function findByRule(
  sourceCode: string,
  yamlRule: string
): Promise<Match[]> {
  const fullRule = yaml.load(yamlRule) as { id: string; rule: object };
  const ruleId = fullRule.id;
  const rulePattern = fullRule.rule;

  const rulePatternAsYaml = yaml.dump(rulePattern);

  const sourceTree = codeParser.parse(sourceCode);
  const intermediateMatches = await ruleProcessor.process(
    sourceTree,
    rulePatternAsYaml
  );

  return intermediateMatches.map(({ node, metavariables }) => {
    const { startLine, endLine } = getLineNumbers(sourceCode, node);
    const matchedText = getMatchText(node);

    return {
      ruleId,
      node,
      matchedText,
      startLine,
      endLine,
      metavariables,
    };
  });
}