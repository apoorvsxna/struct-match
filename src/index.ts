/**
 * This is the main entry point for your library.
 * It exports the public-facing API functions and the core data types.
 */
export * from './api';
export * from './types';