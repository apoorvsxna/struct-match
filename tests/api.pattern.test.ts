import { findByPattern } from '../src/api'; // Changed import
import { patternCases } from './cases/pattern-cases';

describe('findByPattern', () => { // Changed description
  patternCases.forEach(testCase => {
    it(`should correctly handle: ${testCase.name}`, async () => {
      const matches = await findByPattern(testCase.sourceCode, testCase.pattern); // Changed function call
      
      const simplifiedMatches = matches.map(match => ({
        matchedText: match.matchedText,
        metavariables: match.metavariables,
      }));

      expect(simplifiedMatches).toMatchSnapshot();
    });
  });
});