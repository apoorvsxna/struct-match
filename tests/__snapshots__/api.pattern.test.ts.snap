// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`findByPattern should correctly handle: Bare Identifier 1`] = `
[
  {
    "matchedText": "y",
    "metavariables": {},
  },
]
`;

exports[`findByPattern should correctly handle: Chained expressions 1`] = `
[
  {
    "matchedText": "req.body.input",
    "metavariables": {},
  },
]
`;

exports[`findByPattern should correctly handle: Chained expressions with metavariables 1`] = `
[
  {
    "matchedText": "req.body.input",
    "metavariables": {
      "$X": "req",
      "$Y": "body",
      "$Z": "input",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Direct matching 1`] = `
[
  {
    "matchedText": "const x = 5;",
    "metavariables": {},
  },
]
`;

exports[`findByPattern should correctly handle: Metavariable consistency 1`] = `
[
  {
    "matchedText": "const x = x + 5;",
    "metavariables": {
      "$A": "x",
      "$B": "5",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Metavariable matching complex node 1`] = `
[
  {
    "matchedText": "console.log( go(x+5) )",
    "metavariables": {
      "$X": "go(x+5)",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Metavariables 1`] = `
[
  {
    "matchedText": "const x = 5;",
    "metavariables": {
      "$A": "x",
    },
  },
  {
    "matchedText": "const y = 5;",
    "metavariables": {
      "$A": "y",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Metavariables on both sides 1`] = `
[
  {
    "matchedText": "const x = 5;",
    "metavariables": {
      "$A": "x",
      "$B": "5",
    },
  },
  {
    "matchedText": "const y = 10;",
    "metavariables": {
      "$A": "y",
      "$B": "10",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Semicolon 1`] = `
[
  {
    "matchedText": "const x = 5;",
    "metavariables": {},
  },
]
`;

exports[`findByPattern should correctly handle: Varying node type due to context 1`] = `
[
  {
    "matchedText": "{
                        SSL_VERIFYPEER: 0
                    }",
    "metavariables": {},
  },
]
`;

exports[`findByPattern should correctly handle: Whitespace 1`] = `
[
  {
    "matchedText": "const x=  5;",
    "metavariables": {
      "$A": "x",
      "$B": "5",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Wildcard matching multiple nodes 1`] = `
[
  {
    "matchedText": "function go() {
                var a = 5;
                var c = 10;
                if(c == 2) {
                    c = c + 1;
                }
                console.log("random");
            }",
    "metavariables": {
      "$X": ""random"",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Wildcard matching one node 1`] = `
[
  {
    "matchedText": "function go() {
                var a = 5;
                console.log("random");
            }",
    "metavariables": {
      "$X": ""random"",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Wildcard matching zero nodes 1`] = `
[
  {
    "matchedText": "function go() {
                console.log("random");
            }",
    "metavariables": {
      "$X": ""random"",
    },
  },
]
`;

exports[`findByPattern should correctly handle: Wildcard with nesting 1`] = `
[
  {
    "matchedText": "function go(x) {
                var a = 5;
                if(x > a) {
                    x = x+10;
                    console.log("yes");
                }
                return x;
            }",
    "metavariables": {},
  },
]
`;
